collector/dfsr.go

```diff
@@ -0,0 +1,868 @@
+			prometheus.BuildFQName(Namespace, subsystem, "staging_bytes_cleaned_up_total"),
+			"",
+			[]string{"name"},
+			nil,
+		),
+
+		FolderStagingBytesGeneratedTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "staging_bytes_generated_total"),
+			"",
+			[]string{"name"},
+			nil,
+		),
+
+		FolderStagingFilesCleanedUpTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "staging_files_cleaned_up_total"),
+			"",
+			[]string{"name"},
+			nil,
+		),
+
+		FolderStagingFilesGeneratedTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "staging_files_generated_total"),
+			"",
+			[]string{"name"},
+			nil,
+		),
+
+		FolderUpdatesDroppedTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "updates_dropped_total"),
+			"",
+			[]string{"name"},
+			nil,
+		),
+
+		// Volume
+		VolumeDatabaseCommitsTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "database_commits_total"),
+			"Total number of DFSR Volume database commits",
+			[]string{"name"},
+			nil,
+		),
+
+		VolumeDatabaseLookupsTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "database_lookups_total"),
+			"Total number of DFSR Volume database lookups",
+			[]string{"name"},
+			nil,
+		),
+
+		VolumeUSNJournalUnreadPercentage: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "usn_journal_unread_percentage"),
+			"Percentage of DFSR Volume USN journal records that are unread",
+			[]string{"name"},
+			nil,
+		),
+
+		VolumeUSNJournalRecordsAcceptedTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "usn_journal_records_accepted_total"),
+			"Total number of USN journal records accepted",
+			[]string{"name"},
+			nil,
+		),
+
+		VolumeUSNJournalRecordsReadTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "usn_journal_records_read_total"),
+			"Total number of DFSR Volume USN journal records read",
+			[]string{"name"},
+			nil,
+		),
+	}
+
+	dfsrCollector.dfsrChildCollectors = dfsrCollector.getDFSRChildCollectors()
+
+	return &dfsrCollector, nil
+}
+
+// Maps child collectors names to their relevant collection function,
+// for use in DFSRCollector.Collect()
+func (c *DFSRCollector) getDFSRChildCollectors() dfsrCollectorMap {
+	dfsrCollectors := make(dfsrCollectorMap)
+	dfsrCollectors["connection"] = c.collectConnection
+	dfsrCollectors["folder"] = c.collectFolder
+	dfsrCollectors["volume"] = c.collectVolume
+
+	return dfsrCollectors
+}
+
+// Collect implements the Collector interface.
+// Sends metric values for each metric to the provided prometheus Metric channel.
+func (c *DFSRCollector) Collect(ctx *ScrapeContext, ch chan<- prometheus.Metric) error {
+	wg := sync.WaitGroup{}
+
+	for name, function := range c.dfsrChildCollectors {
+		wg.Add(1)
+		go c.execute(ctx, name, function, ch, &wg)
+
+	}
+	wg.Wait()
+
+	if c.dfsrChildCollectorFailure > 0 {
```

Comment:
This looks like it'll keep returning an error if we ever had one, or am I missing a reset somewhere? Regardless, this will be problematic if we have concurrent scrapes.
Since you use perflib in this collector, I'd expect each child to finish very quickly, and we might not need to fan out? Then you could just make a regular call to each child serially and return an error.

Reply:
I think you're correct: there's no reset of this internal counter. This is also the case in the `mssql` collector, so I'll fix that too.
Running the collectors serially would be simpler than keeping the concurrent scrapes and trying to fix with `atomic/sync`, so I'll change that too.